File: solutions/scala/074-binary-tree-maximum-path-sum.scala
/**
 * Problem 74: Binary Tree Maximum Path Sum (LeetCode 124)
 * Difficulty: Hard
 * Language: Scala
 */
def maxPathSum(t: Tree[Int]): Int = {
  def go(node: Tree[Int]): (Int, Int) = node match {
    case Nil => (0, Int.MinValue)
    case Node(v, l, r) =>
      val (lg, lm) = go(l); val (rg, rm) = go(r)
      val gain = 0.max(v + lg.max(rg))
      val pathMax = List(lm, rm, v + lg + rg).max
      (gain, pathMax)
  }
  go(t)._2
}

File: solutions/rust/074-binary-tree-maximum-path-sum.rs
/**
 * Problem 74: Binary Tree Maximum Path Sum (LeetCode 124)
 * Difficulty: Hard
 * Language: Rust
 */
fn max_path_sum(root: &Option<Box<TreeNode>>) -> i64 {
    fn go(node: &Option<Box<TreeNode>>) -> (i64, i64) {
        match node {
            None => (0, i64::MIN),
            Some(n) => {
                let (lg, lm) = go(&n.left);
                let (rg, rm) = go(&n.right);
                let gain = 0i64.max(n.val as i64 + lg.max(rg));
                let path_max = *[lm, rm, n.val as i64 + lg + rg].iter().max().unwrap();
                (gain, path_max)
            }
        }
    }
    go(root).1
}
